crates/subspace-farmer/src/plot.rs

```diff
@@ -376,3 +334,207 @@ impl WeakPlot {
+
+        let piece_offset_to_index_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(base_directory.as_ref().join("plot-offset-to-index.bin"))
+            .map_err(PlotError::OffsetDbOpen)?;
+
+        let piece_index_hash_to_offset_db = IndexHashToOffsetDB::open_default(
+            base_directory.as_ref().join("plot-index-to-offset"),
+        )?;
+
+        Ok(Self {
+            plot,
+            piece_index_hash_to_offset_db,
+            piece_offset_to_index_file,
+            piece_count,
+        })
+    }
+
+    pub fn read_encoding(
+        &mut self,
+        piece_index_hash: PieceIndexHash,
+        mut buffer: impl AsMut<[u8]>,
+    ) -> io::Result<()> {
+        let offset = self
+            .piece_index_hash_to_offset_db
+            .get(piece_index_hash)?
+            .ok_or_else(|| {
+                io::Error::other(format!("Piece with hash {piece_index_hash:?} not found"))
+            })?;
+        self.plot.seek(SeekFrom::Start(offset))?;
+        self.plot.read_exact(buffer.as_mut())
+    }
+
+    fn get_piece_index(&mut self, offset: PieceOffset) -> io::Result<PieceIndex> {
+        let mut buf = [0; 8];
+        self.piece_offset_to_index_file.seek(SeekFrom::Start(
+            offset * std::mem::size_of::<PieceIndex>() as u64,
+        ))?;
+        self.piece_offset_to_index_file.read_exact(&mut buf)?;
+        Ok(PieceIndex::from_le_bytes(buf))
+    }
+
+    fn put_piece_index(&mut self, offset: PieceOffset, piece_index: PieceIndex) -> io::Result<()> {
+        self.piece_offset_to_index_file.seek(SeekFrom::Start(
+            offset * std::mem::size_of::<PieceIndex>() as u64,
+        ))?;
+        self.piece_offset_to_index_file
+            .write_all(&piece_index.to_le_bytes())
+    }
+
+    pub fn do_plot(
+        mut self,
+        any_requests_receiver: mpsc::Receiver<()>,
+        read_requests_receiver: mpsc::Receiver<ReadRequests>,
+        write_requests_receiver: mpsc::Receiver<WriteRequests>,
+    ) {
+        let mut did_nothing = true;
+        loop {
+            if did_nothing {
+                // Wait for stuff to come in
+                if any_requests_receiver.recv().is_err() {
+                    break;
+                }
+            }
+
+            did_nothing = true;
+
+            // Process as many read requests as there is
+            while let Ok(read_request) = read_requests_receiver.try_recv() {
+                did_nothing = false;
+
+                match read_request {
+                    ReadRequests::ReadEncoding {
+                        index_hash,
+                        result_sender,
+                    } => {
+                        let mut buffer = Piece::default();
+                        let result = self.read_encoding(index_hash, &mut buffer).map(|()| buffer);
+                        let _ = result_sender.send(result);
```

Comment:
I agree with @liuchengxu, I don't see how this is better than this:
```suggestion
                        let result = self.read_encoding(index_hash);
                        let _ = result_sender.send(result);
```

The buffer is fixed for one element, doesn't make sense to have a mutable buffer as an argument there. It is trivial to write a piece of code that does exactly that. So I'm wondering why `read_encoding` even needs to exist then. We can always add more abstractions later.